homeassistant/components/enasolar/config_flow.py

```diff
@@ -0,0 +1,202 @@
+"""EnaSolar solar inverter configuration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import pyenasolar
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CAPABILITY,
+    CONF_CAPABILITY,
+    CONF_DC_STRINGS,
+    CONF_HOST,
+    CONF_MAX_OUTPUT,
+    CONF_NAME,
+    CONF_SUN_DOWN,
+    CONF_SUN_UP,
+    DC_STRINGS,
+    DEFAULT_HOST,
+    DEFAULT_NAME,
+    DEFAULT_SUN_DOWN,
+    DEFAULT_SUN_UP,
+    DOMAIN,
+    MAX_OUTPUT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EnaSolarConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """User Configuration of EnaSolar Integration."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._enasolar = pyenasolar.EnaSolar()  # Only exposes methods, no significant
+        # code is executed during instantiation
+        self._data: dict[str, Any] = {}
+        self._errors: dict[str, str] = {}
+
+    def _conf_for_inverter_exists(self, serial) -> bool:
+        """Return True if inverter exists in configuration."""
+        for entry in self._async_current_entries(include_ignore=False):
+            if serial == entry.unique_id:
+                return True
+        return False
+
+    async def _try_connect(self, host):
+        try:
+            await self._enasolar.interogate_inverter(host)
+            return True
+
+        except Exception as conerr:  # pylint: disable=broad-except
+            self._errors[CONF_HOST] = "cannot_connect"
+            _LOGGER.error(
+                "Connection to EnaSolar inverter '%s' failed (%s)", host, conerr
+            )
+        return False
+
+    def get_name(self):
+        """Needed to mock name when running tests."""
+        return self._data[CONF_NAME]
+
+    def get_serial_no(self):
+        """Needed to mock serial_no when running tests."""
+        return self._enasolar.get_serial_no()
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return EnaSolarOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None):
+        """Step when user initializes a integration."""
+        _errors = {}
+
+        if user_input is not None:
+            self._data[CONF_HOST] = user_input[CONF_HOST]
+            self._data[CONF_NAME] = user_input[CONF_NAME]
+            if await self._try_connect(self._data[CONF_HOST]):
+                if self._conf_for_inverter_exists(self.get_serial_no()):
+                    return self.async_abort(reason="already_configured")
+                await self.async_set_unique_id(self.get_serial_no())
+                return await self.async_step_inverter()
+            _errors = self._errors
```

Comment:
The config flow design does not allow passing errors between steps. 

Errors should only apply to the current step.

Reply:
Reorganised to use try instead